lib/bulk/common.js

```diff
@@ -445,44 +454,41 @@ function mergeBatchResults(batch, bulkResult, err, result) {
     return;
   }
 
-  // Deal with opTime if available
+  // The server write command specification states that lastOp is an optional
+  // mongod only field that has a type of timestamp. Across various scare specs
+  // where opTime is mentioned, it is an "opaque" object that can have a "ts" and
+  // "t" field with Timestamp and Long as their types respectively.
+  // The "lastOp" field of the bulk write result is never mentioned in the driver
+  // specifications or the bulk write spec, so we should probably just keep its
+  // value consistent since it seems to vary.
+  // See: https://github.com/mongodb/specifications/blob/master/source/driver-bulk-update.rst#results-object
   if (result.opTime || result.lastOp) {
-    const opTime = result.lastOp || result.opTime;
-    let lastOpTS = null;
-    let lastOpT = null;
-
-    // We have a time stamp
-    if (opTime && opTime._bsontype === 'Timestamp') {
-      if (bulkResult.lastOp == null) {
-        bulkResult.lastOp = opTime;
-      } else if (opTime.greaterThan(bulkResult.lastOp)) {
-        bulkResult.lastOp = opTime;
-      }
-    } else {
-      // Existing TS
-      if (bulkResult.lastOp) {
-        lastOpTS =
-          typeof bulkResult.lastOp.ts === 'number'
-            ? Long.fromNumber(bulkResult.lastOp.ts)
-            : bulkResult.lastOp.ts;
-        lastOpT =
-          typeof bulkResult.lastOp.t === 'number'
-            ? Long.fromNumber(bulkResult.lastOp.t)
-            : bulkResult.lastOp.t;
+    let opTime = result.lastOp || result.opTime;
+
+    if (opTime) {
+      // If the opTime is a Timestamp, convert it to a consistent format to be
+      // able to compare easily. Converting to the object from a timestamp is
+      // much more straightforward than the other direction.
+      if (opTime._bsontype === 'Timestamp') {
```

Comment:
should we have an `else` with an assertion about the shape to make sure that if it's unexpected we fail in a predictable manner?